### ISSUE TITLE
Page weak reference is still alive after pushing and popping pages modally or normally

### ISSUE THREAD

**Issue by MDThomsen** [dotnet/maui#21143]
### Description

Taking inspiration from [this guide from Jonathan Peppers](https://github.com/dotnet/maui/wiki/Memory-Leaks#diagnosing-memory-leaks) I wanted to investigate why I am experiencing memory increases over time when pushing and popping pages both modally and normally. 

I am setting up a weak reference to the page that I am pushing, and then popping. After popping I am performing a GC.Collect(), but I notice that the weak reference is still alive afterwards. I would assume that the weak reference would no longer be alive after popping the page from the view hierarchy.

Code sample:
```
WeakReference pageReference;

{
     var page = new MainPage("New Modal");
     pageReference = new WeakReference(page);

     await Navigation.PushModalAsync(page);
     await Navigation.PopModalAsync();
}

await Task.Yield();
GC.Collect();
GC.WaitForPendingFinalizers();

Console.WriteLine($"Page still alive: {pageReference?.IsAlive}");
```

Otherwise please check out the linked sample repo.

### Steps to Reproduce

1. Create a weak reference to a page and push it normally or modally within it's own scope.
2. Pop the page
3. Call Garbage Collector
4. Check if the weak reference is still alive. (It is)

### Link to public reproduction project repository

https://github.com/MDThomsen/Weak-Reference-MAUI-Sample/

### Version with bug

8.0.6 SR1

### Is this a regression from previous behavior?

No, this is something new

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 14

**Comment by mattleibow** [comment]
@jonathanpeppers thoughts?

**Comment by jonathanpeppers** [comment]
@MDThomsen do can you record a `.gcdump` file of when it has leaked like this?

We just recently revamped Android instructions here:

* https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/tracing.md
* https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/tracing.md#memory-dumps-for-android-in-net-8

**Comment by jonathanpeppers** [comment]
Oh, if the sample uses `TabbedPage` it might be the same as this one:

* https://github.com/dotnet/maui/issues/17959

https://github.com/MDThomsen/Weak-Reference-MAUI-Sample/blob/main/Sample/Sample/Sample/ExtendedTabbedPage.cs

**Comment by MDThomsen** [comment]
@jonathanpeppers Here is the attached gcdump, where the page is getting pushed and popped five times in a row.
[20240313_084013_42800.zip](https://github.com/dotnet/maui/files/14584241/20240313_084013_42800.zip)

Yeah, I forgot to make some more concise names for the pages. The MainPage set in the App is a TabbedPage which has one child which is a ContentPage that pushes and pops a copy of it self.

**Comment by jonathanpeppers** [comment]
@MDThomsen can you try the latest "nightly build"?

* https://github.com/dotnet/maui/wiki/Nightly-Builds

`8.0.20-nightly.10336`, for example has a fix for `TabbedPage` on Android: 23171190fc32ba18c67d54f7b0dd9a3c487e8058
